### ISSUE TITLE
GPU inference time too slow

### ISSUE THREAD

**Issue by peterhon123** [mozilla/DeepSpeech#2292]
For support and discussions, please use our [Discourse forums](https://discourse.mozilla.org/c/deep-speech).

If you've found a bug, or have a feature request, then please create an issue with the following information:

- **Have I written custom code (as opposed to running examples on an unmodified clone of the repository)**: No 
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (our builds, or upstream TensorFlow)**:
- **TensorFlow version (use command below)**: b'v1.13.1-0-g6612da8951' 1.13.1
- **Python version**:  3.6.8
- **Bazel version (if compiling from source)**: Yes( I build DeepSpeech from Dockerfile)
- **GCC/Compiler version (if compiling from source)**:  Yes( I build DeepSpeech from Dockerfile)
- **CUDA/cuDNN version**: CUDA 10 / cnDNN 7
- **GPU model and memory**: TitanX 12212MiB
- **Exact command to reproduce**: 

You can obtain the TensorFlow version with

```bash
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"
```

Please describe the problem clearly. Be sure to convey here why it's a bug or a feature request.

Include any logs or source code that would be helpful to diagnose the problem. For larger logs, link to a Gist, not a screenshot. If including tracebacks, please include the full traceback. Try to provide a reproducible test case.

<hr>

```
root@5015f6babe93:/DeepSpeech# deepspeech --model result/thchs30_model/output_graph.pb --alphabet /home/data/data_thchs30/thchs30_csv/test_collection/thchs30_alphabet.txt --lm /home/data/data_thchs30/thchs30_csv/test_collection/thchs30-l
m.binary --trie /home/data/data_thchs30/thchs30_csv/test_collection/thchs30-trie --audio /home/data/data_thchs30/data/D11_
Display all 498 possibilities? (y or n)
root@5015f6babe93:/DeepSpeech# deepspeech --model result/thchs30_model/output_graph.pb --alphabet /home/data/data_thchs30/thchs30_csv/test_collection/thchs30_alphabet.txt --lm /home/data/data_thchs30/thchs30_csv/test_collection/thchs30-l
m.binary --trie /home/data/data_thchs30/thchs30_csv/test_collection/thchs30-trie --audio /home/data/data_thchs30/data/D11_785.wav
Loading model from file result/thchs30_model/output_graph.pb
TensorFlow: v1.13.1-13-g174b4760eb
DeepSpeech: v0.5.1-0-g4b29b78
Warning: reading entire model file into memory. Transform model file into an mmapped graph to reduce heap usage.
2019-08-07 06:28:31.404612: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-08-07 06:28:31.654287: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1433] Found device 0 with properties:
name: GeForce GTX TITAN X major: 5 minor: 2 memoryClockRate(GHz): 1.076
pciBusID: 0000:04:00.0
totalMemory: 11.93GiB freeMemory: 11.81GiB
2019-08-07 06:28:31.654346: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1483] Ignoring visible gpu device (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:04:00.0, compute capability: 5.2) with Cuda compute capability 5.2. The minimum required Cuda capability is 6.0.
2019-08-07 06:28:31.654370: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-08-07 06:28:31.654384: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990]      0
2019-08-07 06:28:31.654396: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1003] 0:   N
2019-08-07 06:28:31.693160: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "UnwrapDatasetVariant" device_type: "GPU" host_memory_arg: "input_handle" host_memory_arg: "output_handle"') for unknown op: UnwrapDatasetVariant
2019-08-07 06:28:31.693226: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "UnwrapDatasetVariant" device_type: "CPU"') for unknown op: UnwrapDatasetVariant
2019-08-07 06:28:31.693245: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "WrapDatasetVariant" device_type: "GPU" host_memory_arg: "input_handle" host_memory_arg: "output_handle"') for unknown op: WrapDatasetVariant
2019-08-07 06:28:31.693261: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "WrapDatasetVariant" device_type: "CPU"') for unknown op: WrapDatasetVariant
Loaded model in 0.295s.
Loading language model from files /home/data/data_thchs30/thchs30_csv/test_collection/thchs30-lm.binary /home/data/data_thchs30/thchs30_csv/test_collection/thchs30-trie
Loaded language model in 0.0023s.
Running inference.
^CKilled
root@5015f6babe93:/DeepSpeech# deepspeech --model result/thchs30_model/output_graph.pb --alphabet /home/data/data_thchs30/thchs30_csv/test_collection/thchs30_alphabet.txt --lm /home/data/data_thchs30/thchs30_csv/test_collection/thchs30-lm.binary --trie /home/data/data_thchs30/thchs30_csv/test_collection/thchs30-trie --audio /home/data/data_thchs30/data/D11_785.wav
Loading model from file result/thchs30_model/output_graph.pb
TensorFlow: v1.13.1-13-g174b4760eb
DeepSpeech: v0.5.1-0-g4b29b78
Warning: reading entire model file into memory. Transform model file into an mmapped graph to reduce heap usage.
2019-08-07 06:35:08.626432: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-08-07 06:35:08.868645: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1433] Found device 0 with properties:
name: GeForce GTX TITAN X major: 5 minor: 2 memoryClockRate(GHz): 1.076
pciBusID: 0000:04:00.0
totalMemory: 11.93GiB freeMemory: 11.81GiB
2019-08-07 06:35:08.868703: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1483] Ignoring visible gpu device (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:04:00.0, compute capability: 5.2) with Cuda compute capability 5.2.The minimum required Cuda capability is 6.0.
2019-08-07 06:35:08.868728: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-08-07 06:35:08.868742: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990]      0
2019-08-07 06:35:08.868755: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1003] 0:   N
2019-08-07 06:35:08.901570: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "UnwrapDatasetVariant" device_type: "GPU" host_memory_arg: "input_handle" host_memory_arg: "output_handle"') for unknown op: UnwrapDatasetVariant
2019-08-07 06:35:08.901639: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "UnwrapDatasetVariant" device_type: "CPU"') for unknown op: UnwrapDatasetVariant
2019-08-07 06:35:08.901657: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "WrapDatasetVariant" device_type: "GPU" host_memory_arg: "input_handle" host_memory_arg: "output_handle"') for unknown op: WrapDatasetVariant
2019-08-07 06:35:08.901673: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "WrapDatasetVariant" device_type: "CPU"') for unknown op: WrapDatasetVariant
Loaded model in 0.281s.
Loading language model from files /home/data/data_thchs30/thchs30_csv/test_collection/thchs30-lm.binary /home/data/data_thchs30/thchs30_csv/test_collection/thchs30-trie
Loaded language model in 0.00161s.
Running inference.
北 京 丰 台 区 农 民 自 己 发 展 潮 办 万 佛 延 寿 子 迎 春 药 会 需 月 文 区 内 六 十 春 江 更 对 产 赛
Inference took 879.628s for 9.858s audio file.
```

The inference time is almost 100 times the audio duration. May i ask whether something i miss ? It seems in the normal situation it shouldn't act like that.

**Comment by kdavis-mozilla** [comment]
This is likely a reflection of the size of your alphabet.

For Mandarin the alphabet size is usually around 5-8K, 150-250 times the size for English, which make the softmax layer much more computationally intensive which slows the inference down.

**Comment by peterhon123** [comment]
@kdavis-mozilla thanks for reply. I noticed the message
`2019-08-07 06:35:08.868703: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1483] Ignoring visible gpu device (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:04:00.0, compute capability: 5.2) with Cuda compute capability 5.2.The minimum required Cuda capability is 6.0.`

It seems that the GTX TITAN X capability less than 6 therefore the program use cpu to inference.

**Comment by peterhon123** [comment]
@kdavis-mozilla i think you are correct. 

```
2019-08-07 07:25:31.587818: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1512] Adding visible gpu devices: 0
2019-08-07 07:25:31.888516: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-08-07 07:25:31.888545: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990]      0
2019-08-07 07:25:31.888551: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1003] 0:   N
2019-08-07 07:25:31.888635: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10426 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:17:00.0, compute capability: 6.1)
2019-08-07 07:25:31.902141: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "UnwrapDatasetVariant" device_type: "GPU" host_memory_arg: "input_handle" host_memory_arg: "output_handle"') for unknown op: UnwrapDatasetVariant
2019-08-07 07:25:31.902172: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "UnwrapDatasetVariant" device_type: "CPU"') for unknown op: UnwrapDatasetVariant
2019-08-07 07:25:31.902181: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "WrapDatasetVariant" device_type: "GPU" host_memory_arg: "input_handle" host_memory_arg: "output_handle"') for unknown op: WrapDatasetVariant
2019-08-07 07:25:31.902187: E tensorflow/core/framework/op_kernel.cc:1325] OpKernel ('op: "WrapDatasetVariant" device_type: "CPU"') for unknown op: WrapDatasetVariant
Loaded model in 0.403s.
Loading language model from files /home/thchs30_test/thchs30-lm.binary /home/thchs30_test/thchs30-trie
Loaded language model in 0.00249s.
Running inference.
2019-08-07 07:25:32.149073: I tensorflow/stream_executor/dso_loader.cc:152] successfully opened CUDA library libcublas.so.10.0 locally
北 京 丰 台 区 农 民 自 计 划 学 筹 办 万 佛 延 寿 似 迎 春 袅 会 欣 文 区 内 六 十 之 江 个 队 参 赛
Inference took 926.603s for 9.858s audio file.
```

**Comment by reuben** [comment]
Try setting `cutoff_prob` in `StreamingState::processBatch` to a value less than 1. Something like 0.98. It should improve decoding performance significantly, specially for large alphabets.

**Comment by peterhon123** [comment]
@reuben thank you reuben. I tried what you suggest and it seems that the inference time does not change at all. In the next step i would like to try use phoneme to train and intent to reduce the alphabet size.

**Comment by reuben** [comment]
> I tried what you suggest and it seems that the inference time does not change at all.

That is weird. Are you sure you properly recompiled the client with the change?

**Comment by peterhon123** [comment]
@reuben Sorry reuben. That's my fault. I just changed the value but not recompiled the client.

**Comment by wegamekinglc** [comment]
@peterhon123 Does this change really improved your decoding performance?

**Comment by peterhon123** [comment]
@wegamekinglc I haven't tried it yet. Now i am training based on `Pinyin`.
